copycat/api/src/main/java/org/apache/kafka/copycat/data/SchemaProjector.java

```diff
@@ -0,0 +1,209 @@
+        if (writer.isOptional() && !reader.isOptional()) {
+            if (reader.defaultValue() != null) {
+                if (record != null) {
+                    return projectRequiredSchema(writer, reader, record);
+                } else {
+                    return reader.defaultValue();
+                }
+            } else {
+                throw new SchemaProjectorException("Writer schema is optional, however, reader schema does not provide a default value.");
+            }
+        } else {
+            if (record != null) {
+                return projectRequiredSchema(writer, reader, record);
+            }
+            return null;
+        }
+    }
+
+    private static Object projectRequiredSchema(Schema writer, Schema reader, Object record) throws SchemaProjectorException {
+        Object result = null;
+        // handle logical types
+        if (reader.name() != null) {
+            return projectLogical(writer, reader, record);
+        }
+        switch (reader.type()) {
+            case INT8:
+            case INT16:
+            case INT32:
+            case INT64:
+            case FLOAT32:
+            case FLOAT64:
+            case BOOLEAN:
+            case BYTES:
+            case STRING:
+                result = projectPrimitive(writer, reader, record);
+                break;
+            case STRUCT:
+                result = projectStruct(writer, reader, (Struct) record);
+                break;
+            case ARRAY:
+                result = projectArray(writer, reader, record);
+                break;
+            case MAP:
+                result = projectMap(writer, reader, record);
+        }
+        return result;
+    }
+
+    private static Object projectStruct(Schema writer, Schema reader, Struct writerStruct) throws SchemaProjectorException {
+        Struct readerStruct = new Struct(reader);
+        for (Field readerField : reader.fields()) {
+            String fieldName = readerField.name();
+            Field writerField = writer.field(fieldName);
+            if (writerField != null) {
+                Object writerFieldValue = writerStruct.get(fieldName);
+                Object readerFieldValue = project(writerField.schema(), readerField.schema(), writerFieldValue);
+                readerStruct.put(fieldName, readerFieldValue);
+            } else {
+                Object readerDefault;
+                if (readerField.schema().defaultValue() != null) {
+                    readerDefault = readerField.schema().defaultValue();
+                } else {
+                    throw new SchemaProjectorException("Cannot project " + writer.schema() + " to " + reader.schema());
+                }
+                readerStruct.put(fieldName, readerDefault);
+            }
+        }
+        return readerStruct;
+    }
+
+    private static Object projectArray(Schema writer, Schema reader, Object record) throws SchemaProjectorException {
+        if (writer.type() != reader.type()) {
```

Comment:
This is repeated in many of the methods, and I think there are a few other parts of the schema that you want to check are equal. Can we lift this up into `projectRequiredSchema`? I think you also want to check that the name, version, and parameters are equal. Then each of these `projectX` methods only needs to check any pieces that are specific to it, e.g. the key and value schemas.